### ISSUE TITLE
Compatibility issue with 5Dim_battlefield mod

### ISSUE THREAD

**Issue by malomaster** [pixelwhipped/ModMash#26]
Hi.
Error on load when using with 5dim_battlefield mod:
`Error while running setup for entity prototype "hero-turret-1-gun-turret" (ammo-turret): next_update target (5d-gun-turret-02) must have the same fast_replaceable_group (gun-turret != turret)`.
As I know from my little Lua knowledge and some mods tweaking experience, this problem appeared because your gun hero turrets have "gun-turret" as _fast_replaceable_group_, but, if I'm right, all 5Dim's turrets have group just "turrets", so you can easily replace laser turret to sniper turret or small gun tower to tesla tower. And I think the same issue will be with all other your Hero turrets.
Because hero turrets are dynamically generated by script and not hardcoded, so all names also are generated dynamically, I can't just make a little external patch — smth like _data.raw["turret"]["hero-turret"].fast_replaceable_group = "turret"_ in external data-final-fixes file. Also, because I can't start game for now with this nasty error, I don't know, will "hero turrets" mod include turrets from other mods, and even if not, I don't have so much time to try to find out all hero turrets' generated names to include all them in patch. Moreover, this patch will be an outright crutch, only impairing the performance of the mod. The most annoying thing is that I have tried many times to somehow patch entities generated by 5Dim's mods, but for some reason none of the patch methods on startup I tried worked for them. Maybe 5Dim's entities generated on map loading on something like that instead of generating on game loading, I don't know, what matters is that patches for 5Dim's mods just doesn't working.
So can you add modcheck _if mods["5dim_battlefield"] then_ and a patch which will change _fast_replaceable_group_ for hero turrets to "turret" to get rid of this mods conflict? Or will change hero turrets' _next_update_ to _nil_, if that method doesn't work?

**Comment by pixelwhipped** [comment]
Haven't been working on the mod in well over a year, though some still can and may be on discord.  However if I changed the fast replace group it would more than likely just cause issues elsewhere.  I think it would be more a case of just saying not compatible and one or the other mods need be deleted.  Game should start then.I would look at a mod check but not sure I could figure how to do correctly.  I wish I could designate time for it but I'm simply very time constrained these daysSent from my Galaxy
-------- Original message --------From: malomaster ***@***.***> Date: 14/09/23  3:38 pm  (GMT+12:00) To: pixelwhipped/ModMash ***@***.***> Cc: Subscribed ***@***.***> Subject: [pixelwhipped/ModMash] Compatibility issue with 5Dim_battlefield mod (Issue #26) 
Hi.
Error on load when using with 5dim_battlefield mod:
Error while running setup for entity prototype "hero-turret-1-gun-turret" (ammo-turret): next_update target (5d-gun-turret-02) must have the same fast_replaceable_group (gun-turret != turret).
As I know from my little Lua knowledge and some mods tweaking experience, this problem appeared because your gun hero turrets have "gun-turret" as fast_replaceable_group, but, if I'm right, all 5Dim's turrets have group just "turrets", so you can easily replace laser turret to sniper turret or small gun tower to tesla tower. And I think the same issue will be with all other your Hero turrets.
Because hero turrets are dynamically generated by script and not hardcoded, so all names also are generated dynamically, I can't just make a little external patch — smth like data.raw["turret"]["hero-turret"].fast_replaceable_group = "turret" in external data-final-fixes file. Also, because I can't start game for now with this nasty error, I don't know, will "hero turrets" mod include turrets from other mods, and even if not, I don't have so much time to try to find out all hero turrets' generated names to include all them in patch. Moreover, this patch will be an outright crutch, only impairing the performance of the mod. The most annoying thing is that I have tried many times to somehow patch entities generated by 5Dim's mods, but for some reason none of the patch methods on startup I tried worked for them. Maybe 5Dim's entities generated on map loading on something like that instead of generating on game loading, I don't know, what matters is that patches for 5Dim's mods just doesn't working.
So can you add modcheck if mods["5dim_battlefield"] then and a patch which will change fast_replaceable_group for hero turrets to "turret" to get rid of this mods conflict? Or will change hero turrets' next_update to nil, if that method doesn't work?

**Comment by malomaster** [comment]
Yeah, without "5dim_battlefield" hero turrets works. The problem is I don't wanted to delete 5Dim's mods (at least "battlefield", because it added 3 additional turret types for gun and laser turrets, and all turrets including tesla turret now have 10 levels, every level have more range, also I tweaked it, so next levels also have little damage and attack speed), because with "Rampart" mod and some other mods which add new types of enemies, without those towers (especially sniper and laser sniper turrets) the game became almost unplayable because of too strong and too annoying enemies.
So, if you will have free time, can you at least try to look on "Hero turrets + 5Dim_battlefield" combo and what can be done to fix this issue? I'm not in a hurry and I can still play without "Hero turrets", but I'll be happy if this problem someday will be solved.

**Comment by malomaster** [comment]
I think I found the problem.
https://github.com/millerscout/MMD-Factorio-Mods/issues/1#issuecomment-1761174275
When I added "--" at the start of the strings I mentioned, all worked with "xp-for-building" mod. So it should work for "Hero turrets" too.
